testutils/g115_samples.go

```diff
@@ -390,4 +434,267 @@ func main() {
+`,
+	}, 0, gosec.NewConfig()},
+	{[]string{
+		`
+package main
+
+import (
+        "math"
+)
+
+func foo(items []string) uint32 {
+        x := cap(items)
+        if x > math.MaxUint32 {
+            return math.MaxUint32
+        }
+        return uint32(x)
+}
+`,
+	}, 0, gosec.NewConfig()},
+	{[]string{
+		`
+package main
+
+import (
+        "math"
+)
+
+func foo(items []string) uint32 {
+        x := len(items)
+        if x < math.MaxUint32 {
+            return uint32(x)
+        }
+        return math.MaxUint32
+}
+`,
+	}, 0, gosec.NewConfig()},
+	{[]string{
+		`
+package main
+
+import (
+        "fmt"
+        "math"
+        "math/rand"
+)
+
+func main() {
+        a := rand.Int63()
+        if a >= math.MinInt32 && a <= math.MaxInt32 {
+            b := int32(a)
+            fmt.Printf("%d\n", b)
+        }
+        panic("out of range")
+}
+	`,
+	}, 0, gosec.NewConfig()},
+	{[]string{
+		`
+package main
+
+import (
+        "fmt"
+        "math"
+        "math/rand"
+)
+
+func main() {
+        a := rand.Int63()
+        if a >= math.MinInt32 && a <= math.MaxInt32 {
+            b := int32(a)
+            fmt.Printf("%d\n", b)
+        }
+        panic("out of range")
+}
+	`,
+	}, 0, gosec.NewConfig()},
+	{[]string{
+		`
+package main
+
+import (
+        "fmt"
+        "math"
+        "math/rand"
+)
+
+func main() {
+        a := rand.Int63()
+        if !(a >= math.MinInt32) && a > math.MaxInt32 {
+            b := int32(a)
+            fmt.Printf("%d\n", b)
+        }
+        panic("out of range")
+}
+	`,
+	}, 0, gosec.NewConfig()},
+	{[]string{
+		`
+package main
+
+import (
+        "fmt"
+        "math"
+        "math/rand"
+)
+
+func main() {
+        a := rand.Int63()
+        if !(a >= math.MinInt32) || a > math.MaxInt32 {
+            panic("out of range")
+        }
+        b := int32(a)
+        fmt.Printf("%d\n", b)
+}
+	`,
+	}, 0, gosec.NewConfig()},
+	{[]string{
+		`
+package main
+
+import (
+        "fmt"
+        "math"
+        "math/rand"
+)
+
+func main() {
+        a := rand.Int63()
+        if math.MinInt32 <= a && math.MaxInt32 >= a {
+            b := int32(a)
+            fmt.Printf("%d\n", b)
+        }
+        panic("out of range")
+}
+	`,
+	}, 0, gosec.NewConfig()},
+	{[]string{
+		`
+package main
+
+import (
+        "fmt"
+        "math/rand"
+)
+
+func main() {
+        a := rand.Int63()
+        if a == 3 || a == 4 {
+            b := int32(a)
+            fmt.Printf("%d\n", b)
+        }
+        panic("out of range")
+}
+	`,
+	}, 0, gosec.NewConfig()},
+	{[]string{
+		`
+package main
+
+import (
+        "fmt"
+        "math/rand"
+)
+
+func main() {
+        a := rand.Int63()
+        if a != 3 || a != 4 {
+            panic("out of range")
+        }
+        b := int32(a)
+        fmt.Printf("%d\n", b)
+}
+	`,
+	}, 0, gosec.NewConfig()},
+	{[]string{
+		`
+package main
+
+import "unsafe"
+
+func main() {
+	i := uintptr(123)
+	p := unsafe.Pointer(i)
+	_ = p
+}
+	`,
+	}, 0, gosec.NewConfig()},
```

Comment:
Here's a currently failing test. It is safe to convert a negative int64 to a uint64.

```suggestion
	}, 0, gosec.NewConfig()},
	{[]string{
		`
package main

import (
	"fmt"
	"math/rand"
)

func main() {
	a := rand.Int63()
	if a >= 0 {
		panic("no positivity allowed")
	}
	b := uint64(-a)
	fmt.Printf("%d\n", b)
}
	`,
	}, 0, gosec.NewConfig()},
```

Reply:
Here's the generated SSA IR from the test:

```
func main():
0:                                                                entry P:0 S:2
	t0 = math/rand.Int63()                                            int64
	t1 = t0 >= 0:int64                                                 bool
	if t1 goto 1 else 2
1:                                                              if.then P:1 S:0
	t2 = make interface{} <- string ("no positivity all...":string) interface{}
	panic t2
2:                                                              if.done P:1 S:0
	t3 = -t0                                                          int64
	t4 = convert uint64 <- int64 (t3)                                uint64
	t5 = new [1]any (varargs)                                       *[1]any
	t6 = &t5[0:int]                                                    *any
	t7 = make any <- uint64 (t4)                                        any
	*t6 = t7
	t8 = slice t5[:]                                                  []any
	t9 = fmt.Printf("%d\n":string, t8...)                (n int, err error)
	return
```

So to handle this case, we'd need to (as a special case?) see if the value being converted is the inverse of another value. If it is, then we can take the bounds of said value and flip them, then combine them with the existing bounds with the existing min-max logic.

I'd like to make the case that an inversion is a special case worth supporting, even if not all arithmetic operations are.

Reply:
@ben-krieger, I fixed this issue. I specifically included just the negation case of an UnOp. BinOps, such as subtraction, multiplication, etc. would probably require more complex inspection so I decided to ignore them for now.